clifford/numba/_multivector.py

```diff
@@ -0,0 +1,136 @@
+"""
+Numba support for MultiVector objects.
+
+For now, this just supports .value wrapping / unwrapping
+"""
+import numpy as np
+import numba
+from numba.extending import NativeValue
+
+try:
+    # module locations as of numba 0.49.0
+    import numba.np.numpy_support as _numpy_support
+    from numba.core.imputils import impl_ret_borrowed, lower_constant
+    from numba.core import cgutils, types
+except ImportError:
+    # module locations prior to numba 0.49.0
+    import numba.numpy_support as _numpy_support
+    from numba.targets.imputils import impl_ret_borrowed, lower_constant
+    from numba import cgutils, types
+
+from .._multivector import MultiVector
+
+from ._layout import LayoutType
+
+__all__ = ['MultiVectorType']
+
+
+class MultiVectorType(types.Type):
+    def __init__(self, dtype: np.dtype):
+        assert isinstance(dtype, np.dtype)
+        self.dtype = dtype
+        super().__init__(name='MultiVector[{!r}]'.format(numba.from_dtype(dtype)))
+
+    @property
+    def key(self):
+        return self.dtype
+
+    @property
+    def value_type(self):
+        return numba.from_dtype(self.dtype)[:]
+
+    @property
+    def layout_type(self):
+        return LayoutType()
+
+
+# The docs say we should use register a function to determine the numba type
+# with `@numba.extending.typeof_impl.register(MultiVector)`, but this is way
+# too slow (https://github.com/numba/numba/issues/5839). Instead, we use the
+# undocumented `_numba_type_` attribute, and use our own cache. In future
+# this may need to be a weak cache, but for now the objects are tiny anyway.
+_cache = {}
+
+@property
+def _numba_type_(self):
+    dt = self.value.dtype
+    try:
+        return _cache[dt]
+    except KeyError:
+        ret = _cache[dt] = MultiVectorType(dtype=dt)
+        return ret
+
+MultiVector._numba_type_ = _numba_type_
```

Comment:
xref https://github.com/numba/numba/issues/5839

Reply:
@hameerabbasi, you might want to use a similar approach in `sparse`, it gives a sizeable performance boost.

Reply:
Compilation time, call time outside Numba, or call time within Numba? I assume call time outside Numba.

Reply:
The latter, time to enter a numba function that accepts a sparse matrix. Here at Clifford we never actually used the sparse support I wrote, because the performance cost was seemingly too high.